### PR TITLE
2268 update reporting

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #2268.

Updates reporting to use the repositories when building the scan list.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [ ] ~~Was the documentation framework updated to reflect the changes?~~
* [ ] ~~Have you checked that you haven't introduced any duplicate code?~~

## Testing Checklist

* [ ] ~~Added relevant unit tests?~~
* [ ] ~~Have you successfully tested your changes locally? Elaborate:~~
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working
image
image

### CODE REVIEW

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 1]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
+            ]
             formatted_nodes.append(
                 {
-                    "label": node["label"],
-                    "ip_addresses": node["ip_addresses"],
-                    "accessible_from_nodes": nodes_that_can_access_current_node,
-                    "services": node["services"],
-                    "domain_name": node["domain_name"],
-                    "pba_results": node["pba_results"] if "pba_results" in node else "None",
+                    "label": machine.hostname,
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:127]: Might as well just name this `"hostname"` and let the UI build its own label

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 2]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
 
-        nodes = ReportService.get_all_displayed_nodes()
+        machines = ReportService.get_all_machines()
 
-        for node in nodes:
+        for machine in machines:
             # This information should be evident from the map, not sure a table/list is a good way
             # to display it anyways
-            nodes_that_can_access_current_node = node["accessible_from_nodes_hostnames"]
+            addresses = [str(iface.ip) for iface in machine.network_interfaces]
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:121]: why do we need `str(...)`? Won't jsonifying turn it into string?

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 3]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
-        ]
-        nodes_with_monkeys = [
-            NodeService.get_displayed_node_by_id(monkey["_id"], True)
-            for monkey in mongo.db.monkey.find({}, {"_id": 1})
-        ]
-        nodes = nodes_without_monkeys + nodes_with_monkeys
-        return nodes
+    @classmethod
+    def get_accessible_machines(cls, machine: Machine):
+        if cls._node_repository is None or cls._machine_repository is None:
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:142]: When would this happen? Shouldn't this be an error?

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 4]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
+            return []
+
+        nodes = cls._node_repository.get_nodes()
+        machine_iter = (node for node in nodes if node.machine_id == machine.id)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:146]: Just get nodes by machine ID? But the relationship between node and machine shoud be 1 to 1.

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 5]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
-            for node in mongo.db.node.find({}, {"_id": 1})
-        ]
-        nodes_with_monkeys = [
-            NodeService.get_displayed_node_by_id(monkey["_id"], True)
-            for monkey in mongo.db.monkey.find({}, {"_id": 1})
-        ]
-        nodes = nodes_without_monkeys + nodes_with_monkeys
-        return nodes
+    @classmethod
+    def get_accessible_machines(cls, machine: Machine):
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:141]: accessible != scanned. If a machine A communicated to the island it means that for machine A, the island is accessible. It doesn't mean that machine A scanned the island

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 6]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
+        if cls._node_repository is None or cls._machine_repository is None:
+            return []
+
+        nodes = cls._node_repository.get_nodes()
+        machine_iter = (node for node in nodes if node.machine_id == machine.id)
+        accessible_machines = set()
+        for source in machine_iter:
+            for dest, conn in source.connections.items():
+                if CommunicationType.SCANNED in conn:
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:150]: For scanned it's scanned AND exploited (we can't exploit without scanning?)

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 7]
```diff
@@ -102,39 +113,56 @@ def get_tunnels():
...
+    def get_all_machines(cls) -> Iterable[Machine]:
+        if cls._machine_repository is None:
+            return iter(())
+        machines = cls._machine_repository.get_machines()
+        t1, t2 = tee(machines)
+
+        def is_island(machine: Machine):
+            return machine.island
+
+        return chain(filter(is_island, t1), filterfalse(is_island, t2))
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:165]: So we duplicate the list of machine, then on one list we only leave the islands, on another one we only leave the non-islands. Then we concatenate these lists. The end result is the list we had in the first place?

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 2 above)
Comment (cakekoa) [reply monkey/monkey_island/cc/services/reporting/report.py:121]: Looks like the answer is no: `TypeError: Object of type IPv4Address is not JSON serializable`

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 4 above)
Comment (cakekoa) [reply monkey/monkey_island/cc/services/reporting/report.py:146]: I'm not sure what you mean here, but we could shift this to a `get_nodes_for_machine(machine_id)` method on the node repository

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 7 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/services/reporting/report.py:167]: By "duplicate" I meant deep copy, not code duplication. I've tested this method and I didn't see the point. 1. There's no indication that the intent of this method is to sort machines by `is_island`. 2. If it's a sort method, then why not use sort function? 3. Sorting data is responsibility of the display (UI).

**Comment by VakarisZ** [comment]
The screenshots are not related to the issue. Map is not generated from this, nor are "breached servers". "Scanned servers" is what we should look at

File: monkey/common/agent_event_serializers/register.py [hunk 8]
```diff
@@ -9,3 +9,4 @@ def register_common_agent_event_serializers(
     event_serializer_registry[CredentialsStolenEvent] = PydanticAgentEventSerializer(
         CredentialsStolenEvent
     )
+    event_serializer_registry[PingScanEvent] = PydanticAgentEventSerializer(PingScanEvent)
```
Comment (VakarisZ) [inline monkey/common/agent_event_serializers/register.py:12]: There should be a `PydanticAgentEventSerializerRegistry` which would serialize all events with `PydanticAgentEventSerializer` by default. There's no point in having to manually register each event

Comment (mssalvatore) [reply monkey/common/agent_event_serializers/register.py:12]: Sounds good. Create an issue.

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 9]
```diff
@@ -102,39 +112,44 @@ def get_tunnels():
...
+        machines = ReportService._machine_repository.get_machines()
 
-        logger.info("Scanned nodes generated for reporting")
+        for machine in machines:
+            accessible_from = ReportService.get_scanners_of_machine(machine)
+            if accessible_from:
+                formatted_nodes.append(
+                    {
+                        "hostname": machine.hostname,
+                        "ip_addresses": list(map(str, machine.network_interfaces)),
```
Comment (cakekoa) [inline monkey/monkey_island/cc/services/reporting/report.py:123]: Should this be using the IPs and not the network interfaces? Or should we change the key to `network_interfaces`?

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 10]
```diff
@@ -102,39 +112,44 @@ def get_tunnels():
...
+                        "accessible_from_nodes": list(
+                            map(lambda m: m.dict(simplify=True), accessible_from)
```
Comment (cakekoa) [inline monkey/monkey_island/cc/services/reporting/report.py:125]: `[m.dict(simplify=True) for m in accessible_from]` is more concise, no?

File: monkey/monkey_island/cc/ui/src/components/report-components/security/ScannedServers.js [hunk 11]
```diff
@@ -1,19 +1,20 @@
...
-import {renderArray, renderIpAddresses} from '../common/RenderArrays';
+import {renderArray, renderIpAddresses, renderMachineArray} from '../common/RenderArrays';
 
 
 const columns = [
   {
     Header: 'Scanned Servers',
     columns: [
-      {Header: 'Machine', accessor: 'label'},
+      {Header: 'Machine', id: 'machine', accessor: x => x.ip_addresses[0]},
```
Comment (cakekoa) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/ScannedServers.js:11]: Is the `hostname` not useful here?

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py [hunk 12]
```diff
@@ -232,6 +232,25 @@ def test_get_machine_by_ip__retrieval_error(error_raising_machine_repository):
         error_raising_machine_repository.get_machines_by_ip("1.1.1.1")
 
 
+def test_get_machines(machine_repository):
+    retrieved_machines = machine_repository.get_machines()
+
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py:237]: Also check that lengths are equal. As written, this won't test if `retrieved_machines` has extra machines in it.

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py [hunk 13]
```diff
@@ -232,6 +232,25 @@ def test_get_machine_by_ip__retrieval_error(error_raising_machine_repository):
...
+    for machine in MACHINES:
+        assert machine in retrieved_machines
+
+
+def test_get_machines__empty_repository(machine_repository):
+    machine_repository.reset()
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py:243]: `test_get_new_id__new_id_for_empty_repo` also needs an empty repository. Add an `empty_machine_repository` fixture and use that for both tests.

File: monkey/common/agent_event_serializers/register.py (hunk 8 above)
Comment (VakarisZ) [reply monkey/common/agent_event_serializers/register.py:12]: https://github.com/guardicore/monkey/issues/2378

File: monkey/monkey_island/cc/ui/src/components/report-components/security/ScannedServers.js (hunk 11 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/ui/src/components/report-components/security/ScannedServers.js:11]: Often it's `""`

**Review by mssalvatore — APPROVED** [review]
Small changes to unit tests, but not critical.

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py (hunk 12 above)
Comment (VakarisZ) [reply monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_machine_repository.py:237]: Can't we just assert the two lists?

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 10 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/reporting/report.py:125]: +1

**Pull request merged**